core/state/state_object.go

```diff
@@ -280,7 +281,11 @@ func (c *stateObject) ReturnGas(gas *big.Int) {}
 
 func (self *stateObject) deepCopy(db *StateDB, onDirty func(addr common.Address)) *stateObject {
 	stateObject := newObject(db, self.address, self.data, onDirty)
-	stateObject.trie = self.trie
+	if self.trie != nil {
+		// A shallow copy makes the two tries independent.
+		cpy := *self.trie
+		stateObject.trie = &cpy
```

Comment:
Shallow copy support for SecureTrie was added in [710435b51](https://github.com/ethereum/go-ethereum/commit/710435b51b97b4c688b70bda35ab9d1aa704a988#diff-6ed2f15666b9ea78bab176eaa5c69ba6R202), by yours truly.